src/lib/datatypes/sol-vector.c

```diff
@@ -349,18 +349,20 @@ sol_vector_del_element(struct sol_vector *v, const void *elem)
 SOL_API int
 sol_ptr_vector_del_element(struct sol_ptr_vector *pv, const void *elem)
 {
-    int r;
-    uint16_t i, removed = 0;
-    void *cur;
-
-    SOL_PTR_VECTOR_FOREACH_REVERSE_IDX (pv, cur, i)
-        if (cur == elem) {
-            r = sol_ptr_vector_del(pv, i);
-            SOL_INT_CHECK(r, < 0, r);
+    uint16_t i, removed = 0, offset = 0;
+    void **v = pv->base.data;
+
+    for (i = 0; i < pv->base.len; i++)
+        if (v[i] == elem)
             removed++;
-        }
+        else
+            v[offset++] = v[i];
```

Comment:
branch is always cheaper than memory write, which may invalidate caches and so on.

I did graphics blitting functions before, it did a huge difference if you'd check "alpha == 0" and just skip the bit instead of copying the value to itself.

While at that, what used to help me a lot in those graphics primitives were loop unrolling, that would reduce the impact of the `i < pv->base.len` and `i++`. I used to do 16-entries block and then finish with the single entry for for the remaining parts.